exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,434 @@
+ */
+
+use exonum::{
+    api::ApiContext,
+    blockchain::Schema as CoreSchema,
+    crypto::{Hash, PublicKey, SecretKey},
+    helpers::ValidatorId,
+    messages::BinaryValue,
+    exonum_merkledb::{self, Fork, Snapshot},
+    node::ApiSender,
+    proto::Any,
+    runtime::{
+        api::ServiceApiBuilder,
+        dispatcher::{Dispatcher, DispatcherSender},
+        ArtifactId, ArtifactProtobufSpec, CallInfo, ErrorKind, ExecutionContext, ExecutionError,
+        InstanceDescriptor, InstanceId, InstanceSpec, Runtime, RuntimeIdentifier,
+        StateHashAggregator,
+    },
+};
+use futures::{Future, IntoFuture};
+use jni::{
+    objects::{GlobalRef, JObject, JValue},
+    signature::{JavaType, Primitive},
+    Executor,
+};
+use proto;
+use proxy::node::NodeContext;
+use std::fmt;
+use std::str::FromStr;
+use storage::View;
+use to_handle;
+use utils::{jni_cache::runtime_adapter, panic_on_exception, unwrap_jni};
+use JniErrorKind;
+use JniResult;
+
+/// A proxy for `ServiceRuntimeAdapter`s.
+#[derive(Clone)]
+pub struct JavaRuntimeProxy {
+    exec: Executor,
+    runtime_adapter: GlobalRef,
+}
+
+/// Artifact identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct JavaArtifactId(String);
+
+/// List of possible Java runtime errors.
+#[derive(Debug, Copy, Clone)]
+#[repr(u8)]
+pub enum Error {
+    /// Unable to parse artifact identifier or specified artifact has non-empty spec.
+    IncorrectArtifactId = 0,
+    /// Checked java exception is occurred
+    JavaException = 1,
+    /// Unspecified error
+    UnspecifiedError = 2,
+    /// Not supported operation
+    NotSupportedOperation = 3,
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceStateHashes")]
+struct ServiceStateHashes {
+    instance_id: u32,
+    state_hashes: Vec<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceRuntimeStateHashes")]
+struct ServiceRuntimeStateHashes {
+    runtime_state_hashes: Vec<Vec<u8>>,
+    service_state_hashes: Vec<ServiceStateHashes>,
+}
+
+impl JavaRuntimeProxy {
+    /// Runtime Identifier
+    pub const RUNTIME_ID: RuntimeIdentifier = RuntimeIdentifier::Java;
+
+    /// Creates new `JavaRuntimeProxy` for given `ServiceRuntimeAdapter` object
+    pub fn new(executor: Executor, adapter: GlobalRef) -> Self {
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            Err(Error::IncorrectArtifactId.into())
+        } else {
+            artifact
+                .name
+                .parse()
+                .map_err(|_| Error::IncorrectArtifactId.into())
+        }
+    }
+
+    fn parse_jni<T>(res: JniResult<T>) -> Result<T, ExecutionError> {
```

Comment:
This function hides error description. Can we avoid that?

Reply:
fixed at https://github.com/exonum/exonum-java-binding/pull/1129/commits/17a658560ef4f58bcc3359d4a8251b90002da62f